serverless/serverless
# Integrating Components

## What did you implement:
Integrates Serverless Components with the Serverless Framework.



## How did you implement it:
Integrated with Serverless Components by simply depending on the components cli as a dependency and using its methods (I've kept all the required logic over there to keep things separate). Since components require at least Node 8, we are only integrating components for Node 8+ users.

One conflicting issue I've found is: what happens when there's an empty directory? I know @medikoo is working on an interactive `serverless` command, but the components project also ships with such a functionality. I think what we'll need to do is add support for components templates in @medikoo implementation, although I wish we could separate concerns. For now, the components check just moves on if it's an empty directory so that @medikoo updates kick in.



## How can we verify it:

1. run `serverless create -t aws-nodejs` - this should create a framework project
2. run `serverless deploy` to deploy the newly created project
3. delete all files from cwd
4. in the same cwd, create a components `serverless.yml` file. For a minimal example

```yml

myBucket:
  component: "@serverless/aws-s3"
```
5. run` serverless` again in the same cwd. This should run components instead of the framework as it detected that it's a components project



## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** YES  
**_Is it a breaking change?:_** NO
